core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -885,13 +885,25 @@ public CompletionStats completionStats(String... fields) {
     }
 
     /**
-     * Rolls the tranlog generation.
+     * checks and removes translog files that no longer need to be retained. See
+     * {@link org.elasticsearch.index.translog.TranslogDeletionPolicy} for details
+     */
+    public void trimTranslog() {
+        verifyNotClosed();
+        final Engine engine = getEngine();
+        engine.trimTranslog();
+    }
+
+    /**
+     * Rolls the tranlog generation and cleans unneeded.
      *
      * @throws IOException if any file operations on the translog throw an I/O exception
      */
-    private void rollTranslogGeneration() throws IOException {
+    private void rollTranslogAndTrimGeneration() throws IOException {
         final Engine engine = getEngine();
-        engine.getTranslog().rollGeneration();
+        final Translog translog = engine.getTranslog();
+        translog.rollGeneration();
+        translog.trimUnreferencedReaders();
```

Comment:
hmm you run `translog.trimUnreferencedReaders();` under the engine lock [here](https://github.com/elastic/elasticsearch/pull/25294/files#diff-74a25d57bbb52202a04fe43d8632172eR1321)  but you run it without a lock here so I wonder when we have to run under lock and if we can assert on that?